group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java

```diff
@@ -0,0 +1,327 @@
+
+    /**
+     * The subscription metadata.
+     */
+    private Map<String, TopicMetadata> subscriptionMetadata = Collections.emptyMap();
+
+    /**
+     * The existing target assignment.
+     */
+    private Map<String, Assignment> assignments = Collections.emptyMap();
+
+    /**
+     * The members which have been updated or deleted. Deleted members
+     * are signaled by a null value.
+     */
+    private final Map<String, ConsumerGroupMember> updatedMembers = new HashMap<>();
+
+    /**
+     * Constructs the object.
+     *
+     * @param groupId       The group id.
+     * @param groupEpoch    The group epoch to compute a target assignment for.
+     * @param assignor      The assignor to use to compute the target assignment.
+     */
+    public TargetAssignmentBuilder(
+        String groupId,
+        int groupEpoch,
+        PartitionAssignor assignor
+    ) {
+        this.groupId = Objects.requireNonNull(groupId);
+        this.groupEpoch = groupEpoch;
+        this.assignor = Objects.requireNonNull(assignor);
+    }
+
+    /**
+     * Adds all the existing members.
+     *
+     * @param members   The existing members in the consumer group.
+     * @return This object.
+     */
+    public TargetAssignmentBuilder withMembers(
+        Map<String, ConsumerGroupMember> members
+    ) {
+        this.members = members;
+        return this;
+    }
+
+    /**
+     * Adds the subscription metadata to use.
+     *
+     * @param subscriptionMetadata  The subscription metadata.
+     * @return This object.
+     */
+    public TargetAssignmentBuilder withSubscriptionMetadata(
+        Map<String, TopicMetadata> subscriptionMetadata
+    ) {
+        this.subscriptionMetadata = subscriptionMetadata;
+        return this;
+    }
+
+    /**
+     * Adds the existing target assignments.
+     *
+     * @param assignments   The existing target assignments.
+     * @return This object.
+     */
+    public TargetAssignmentBuilder withTargetAssignments(
+        Map<String, Assignment> assignments
+    ) {
+        this.assignments = assignments;
+        return this;
+    }
+
+    /**
+     * Adds or updates a member. This is useful when the updated member is
+     * not yet materialized in memory.
+     *
+     * @param memberId  The member id.
+     * @param member    The member to add or update.
+     * @return This object.
+     */
+    public TargetAssignmentBuilder addOrUpdateMember(
+        String memberId,
+        ConsumerGroupMember member
+    ) {
+        this.updatedMembers.put(memberId, member);
+        return this;
+    }
+
+    /**
+     * Removes a member. This is useful when the removed member
+     * is not yet materialized in memory.
+     *
+     * @param memberId The member id.
+     * @return This object.
+     */
+    public TargetAssignmentBuilder removeMember(
+        String memberId
+    ) {
+        return addOrUpdateMember(memberId, null);
+    }
+
+    /**
+     * Builds the new target assignment.
```

Comment:
It may be even better to use `target assignment` in this class as it is means the target assignment "for the group".

Reply:
i'm wondering if we should use rename Assignment to MemberAssignment to avoid confusion

Reply:
Yeah, I was thinking about this as well but I did not because we already have a MemberAssignor in the assignor package. That causes issues when importing both of them. Overall, I think that Assignment is fine because it is often used in a Map<String, Assignment> so it is clear that it is per member.